atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketAssignerTest.java

```diff
@@ -0,0 +1,474 @@
+
+        assertBucketState(BucketState.INCOMPLETE);
+        verify(bucketWriter).writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, 15));
+        stateTable.assertThatStateMachineTransitionedThroughTo(
+                BucketStateAndIdentifier.of(2, BucketAssignerState.start(15)),
+                BucketStateAndIdentifier.of(2, BucketAssignerState.opening(15)),
+                BucketStateAndIdentifier.of(2, BucketAssignerState.waitingUntilCloseable(15)));
+    }
+
+    @Test
+    public void immediatelyClosingBubblesUpBucketWriteExceptionWithNoTransition() {
+        stateTable.setInitialState(1, BucketAssignerState.immediatelyClosing(1, 15));
+        setPunchTable(1, 15);
+
+        when(bucketWriter.writeToAllBuckets(1, Optional.empty(), TimestampRange.of(1, 15)))
+                .thenThrow(RUNTIME_EXCEPTION);
+
+        assertThatThrownBy(bucketAssigner::run).isEqualTo(RUNTIME_EXCEPTION);
+        stateTable.assertThatStateMachineTransitionedThroughTo(); // No transition
+    }
+
+    @Test
+    public void createsBucketsUpToMaxIterationsInOneRunIfAllCloseable() {
+        stateTable.setInitialState(1, BucketAssignerState.start(1));
+        setPunchTable(1, 3);
+        // We don't actually care about the start timestamp, since we always use the last end timestamp.
+        setPunchTableForStartWallClockTime(clock.instant().plus(Duration.ofMinutes(10)), 3, 25);
+        setPunchTableForStartWallClockTime(clock.instant().plus(Duration.ofMinutes(20)), 25, 41);
+        setPunchTableForStartWallClockTime(clock.instant().plus(Duration.ofMinutes(30)), 41, 114);
+        setPunchTableForStartWallClockTime(clock.instant().plus(Duration.ofMinutes(40)), 114, 221);
+
+        // The following iteration will not run, as it's the 6th.
+        setPunchTableForStartWallClockTime(clock.instant().plus(Duration.ofMinutes(50)), 221, 420);
+
+        clock.advance(Duration.ofHours(1));
+
+        bucketAssigner.run();
+        stateTable.assertThatStateMachineTransitionedThroughTo(
+                BucketStateAndIdentifier.of(1, BucketAssignerState.immediatelyClosing(1, 3)),
+                BucketStateAndIdentifier.of(2, BucketAssignerState.start(3)),
+                BucketStateAndIdentifier.of(2, BucketAssignerState.immediatelyClosing(3, 25)),
+                BucketStateAndIdentifier.of(3, BucketAssignerState.start(25)),
+                BucketStateAndIdentifier.of(3, BucketAssignerState.immediatelyClosing(25, 41)),
+                BucketStateAndIdentifier.of(4, BucketAssignerState.start(41)),
+                BucketStateAndIdentifier.of(4, BucketAssignerState.immediatelyClosing(41, 114)),
+                BucketStateAndIdentifier.of(5, BucketAssignerState.start(114)),
+                BucketStateAndIdentifier.of(5, BucketAssignerState.immediatelyClosing(114, 221)),
+                BucketStateAndIdentifier.of(6, BucketAssignerState.start(221)));
+        verify(metrics).progressedToBucketIdentifier(5); // We haven't entered 6 yet.
+    }
```

Comment:
maybe processing a backlog that of immediate closes into a regular one could be good!